### PR TITLE
(feat) add spinner and lock browser during save process

### PR DESCRIPTION
**Pull request by Offem-Peter**
### :pushpin: References

* **Issue:** Closes [[https://app.clickup.com/t/2er5du8](https://app.clickup.com/t/2er5du8)] #

### CODE REVIEW

**Comment by tokland** [comment]
I've set the PR as fraft and removed the reviewers, as @adrianq has asked for some changes. @Offem-Peter, set "ready for review" and re-add us as reviewerers when the task is completed, thx!

File: src/components/steps/save/SaveStep.tsx [hunk 1]
```diff
@@ -57,18 +58,21 @@ const SaveStep: React.FC<StepProps> = ({ project, onCancel, action }) => {
     );
 
     const checkExistingDataAndSave = React.useCallback(async () => {
```
Comment (tokland) [inline src/components/steps/save/SaveStep.tsx:60]: Check the linting warnings (shown on `yarn start` output or the JS console or explit `yarn lint`):

```Line 79:8:  React Hook React.useCallback has a missing dependency: 'loading'. Either include it or remove the dependency array```

File: src/components/steps/save/SaveStep.tsx [hunk 2]
```diff
@@ -57,18 +58,21 @@ const SaveStep: React.FC<StepProps> = ({ project, onCancel, action }) => {
...
         }
 
+        loading.updateMessage(i18n.t("Checking Existing Data"));
         const existingDataCheck = await project.checkExistingDataForDataElementsToBeRemoved();
 
+        loading.updateMessage(i18n.t("Saving Project"));
         switch (existingDataCheck.type) {
             case "no-values":
-                return save();
+                return save(loading);
```
Comment (tokland) [inline src/components/steps/save/SaveStep.tsx:75]: There is another save in `saveAndSendExistingDataMessage`, use loading also there (thus, the `loading` argument of `save`  won't be optional)

File: src/components/steps/save/SaveStep.tsx [hunk 3]
```diff
@@ -214,43 +216,52 @@ function useSave(project: Project, action: StepProps["action"], projectInfo: Rea
...
+                    setErrorMessage(JSON.stringify({ response, payload }, null, 2));
+                }
+                loading?.hide();
+            } catch (err: any) {
+                setSaving(false);
+                console.error(err);
+                snackbar.error(err.message || err.toString());
             }
-        } catch (err: any) {
-            setSaving(false);
```
Comment (tokland) [inline src/components/steps/save/SaveStep.tsx:237]: If there is an error, `loading` should be also hidden.

File: src/components/steps/save/SaveStep.tsx [hunk 4]
```diff
@@ -214,43 +216,52 @@ function useSave(project: Project, action: StepProps["action"], projectInfo: Rea
...
-                    action === "create" ? i18n.t("Project created") : i18n.t("Project updated");
-                const msg = `${baseMsg}: ${projectSaved.name}`;
-                history.push(generateUrl("projects"));
-                if (isDev) saveDataValues(api, projectSaved);
-                snackbar.success(msg);
-            } else {
-                setErrorMessage(JSON.stringify({ response, payload }, null, 2));
+    const save = React.useCallback(
+        async (loading?: LoadingState) => {
+            try {
```
Comment (tokland) [inline src/components/steps/save/SaveStep.tsx:221]: You can add a loadin.show here in case the caller did not call it.

**Review by tokland — CHANGES_REQUESTED** [review]

**Comment by Offem-Peter** [comment]
@tokland changes implemented.

**Comment by tokland** [comment]
Ok. Peter, you ask for a re-review with the github interface, the sync button:

image

**Review by tokland — APPROVED** [review]
Just a couple of details I committed, otherwise, all good. Changes I made:

- The argument `loading` of save is not necessary, as the `useSave` hook can get the variable itself.
- Note that after `existingDataCheck`, if the result was "with-values", we need to hide the loading mask so the user can interact with the existing-data-dialog.

**Pull request merged**